### ISSUE TITLE
Simplify translation creation, access and internals.

### ISSUE THREAD

**Issue by Chicken-Bones** [tModLoader/tModLoader#1365]
### What does this proposal attempt to solve or improve?
1. Declaring a translation is more difficult than it needs to be
  a. Should be able to just declare a field/property of type `Translation` and have it autofilled
  b. At the least, `ModType` could have an extension method for getting a translation which starts with `"{Mod.Name}.{GetType().Name}.{Name}"`
  c. All `ModType` subclasses should probably come with a translation key for display name
2. Translation keys are verbose and inconsistent
  a. `Mods.ExampleMod.ItemName.ExampleItem` could be `ExampleMod.ModItem.ExampleItem` and 
  b. `Mods.ExampleMod.ItemTooltip.ExampleItem` could be `ExampleMod.ModItem.ExampleItem.Tooltip`
3. Translating a key is more difficult than it needs to be
4. Using a translation from vanilla, tML or another mod for an item tooltip is difficult
5. Translations lack the ability to 'bind' values (such as current damage), instead, the site of the translation must pass in the values and all translations must follow the same format.
6. Producing a template lang file for a mod (or filling in missing translation keys) is more difficult than it needs to be

### How do we solve all of these?
To be designed! Input desired.

**Comment by Mirsario** [comment]
The `Mods.` prefix on translation keys should stay, unless we're forcing all vanilla keys to be moved to `Terraria.*`? I don't remember that.

**Comment by sgkoishi** [comment]
Personally, I'm planning to host a [Mozilla's Pontoon](https://github.com/mozilla/pontoon) for my Localizer, which is using reflection and IL-tricks to export/import translations during runtime - and actually, I don't care if the key is changing or not (they should be easy to adopt). 

Several points:
* Split translation and codebase
  * If nobody maintains the translation, players will see legacy outdated text, which is misleading - not better than take a screenshot and go for machine translating
  * Complex string composition becomes more complex
    * For example, accessories with hundreds of effects, dynamically rolling. It is more or less a burden to many mod authors.
  * It's also somehow confusing when the player download a 40MB update and the changelog said "We are just updating some foreign translations not related to you", or "To fix a typo in translation, all players were prompted to download the mod again"
* Using cross-mod translation should be not too hard.
Similar tricks are already commonly used in vanilla, `{$ItemName.xxx}`
* Use similar schema for translatable entries
  * For example, all buff/item/npc are `ModTranslation DisplayName` + `ModTranslation Description`, but `MapEntry` is `LocalizedText name` + `ModTranslation translation`

**Comment by Soaryn** [comment]
Having a had a conversation with @JavidPack one of the hassles for mods providing localizations should not be the structure. Allowing mods to be able to, at any point of the load choose what the key should be, and receive a resultant by the time the player gets to the main menu should be feasible without imposing a set "standard" structure. sometimes there may be a better structure for a mod that doesn't fit the set mold.

One concept is to add to `ModTranslation.cs` a `Defer(string key)` method and `IsDeferred` bool flag. This would hold the localization key that is desired until the normal pass through on `ModItem.cs` or similar runs over it for Localizations.

### Example
The modder could defer a key for localizing.
```cs
public override void SetStaticDefaults() {
    DisplayName.Defer($"Mods.ExampleMod.Items.ExampleItem.Name");
}
```

Then, when `ModItem.cs` (as an example) runs over it, it could see that there is a deferred value and use that for Translation
Current implentation:
```cs
this.DisplayName = LocalizationLoader.GetOrCreateTranslation(this.Mod, "ItemName." + this.Name);
//Line 69 in ModItem.cs
```
Would become something like
```cs
if(this.DisplayName.IsDeferred)
    this.DisplayName = LocalizationLoader.GetOrCreateTranslation(DisplayName.Key); : 
else
    this.DisplayNameLocalizationLoader.GetOrCreateTranslation(this.Mod, "ItemName." + this.Name);
```

**Comment by JavidPack** [comment]
Another usage for the above idea is sharing a tooltip. 

For example, imagine 10 items share a similar behavior, and as such all have the same tooltip. In our current situation, there would have to be 10 entries in the localization files for these items. Sure, each of these localization entries could be `{$Mods.ExampleMod.Common.PaperAirplane}`, pointing to the same actual translation in the localization file, but that would add clutter to the localization files. Having each item in code specify that they actually use an existing translation key would also reduce the number of translation keys that are loaded into the dictionaries. The translation files wouldn't need all the dummy entries for each individual item.

**Comment by Chicken-Bones** [comment]
I am 100% in agreement on allowing translatable strings to have the same key, and modders to provide keys at any point, with tML auto key calculation as a convenience fallback. This discussion is likely to rapidly become too long for github, so lets move to a #collaborators thread in Discord, and post summaries here.

**Comment by Chicken-Bones** [comment]
Solved by #3101
